core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -469,8 +486,10 @@ private[spark] class TaskSetManager(
           extraResources,
           serializedTask)
       }
+      (taskDescription,
+        taskDescription.isEmpty && maxLocality == TaskLocality.ANY && pendingTasks.all.nonEmpty)
```

Comment:
nit split this  into different val to be more readable

Reply:
👍 
What are your thoughts on:
I'm thinking instead of making assumptions about what `taskDescription.isEmpty` means, maybe it'd be better to pass maxLocality into `dequeueTask` and then change its logic near the bottom to be something like:

```
    if (TaskLocality.isAllowed(allowedLocality, TaskLocality.ANY)) {
      dequeue(pendingTaskSetToUse.all).foreach { index =>
        return Some((index, TaskLocality.ANY, speculative))
      }
    } else if (maxLocality == TaskLocality.ANY && pendingTasks.all.nonEmpty) {
      hasReject = true
    }
```

Reply:
actually that reminds me of speculative tasks and we need to account for those.

I don't think it hurts to taskDescription.isEmpty means no task was assigned.
if you move it down, it can't be an else if though it would just be a separate if before None is returned. It might make a little more sense down in dequeueTask because its dealing with the localities and choosing which pendingTasks to look at.

Reply:
maybe I'll leave it as is. The old and current impl both don't distinguish between speculative vs not.

Reply:
actually you can't. You are adding in a check for pendingTasks.all.nonEmpty and assuming that is all pending tasks. This is new code and wasn't previously there.  You need to look at both pendingTasks and pendingSpeculatableTasks

Reply:
Thanks for pointing that out